### PR TITLE
[GNA] Issue 39975 - cascade concat fix

### PR DESCRIPTION
**Pull request by kmagiers**

### CODE REVIEW

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 1]
```diff
@@ -454,16 +455,17 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
         static std::map<std::string, size_t> restarted_counter;
         auto restartedCountIt = restarted_counter.find(concatLayer->name);
         if (restartedCountIt == restarted_counter.end()) {
-            auto pos = restarted_counter.insert({ "concatLayer->name", 0 });
+            auto pos = restarted_counter.insert({ concatLayer->name, 0 });
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:458]: most likely it was bug, no comment here

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 2]
```diff
@@ -443,8 +443,9 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
         }
 
         // find a source quant value
-        // - 1st candidate - non-activation layer with non-1 scale factor
-        // - 2nd candidate - 1st layer with non-1 scale factor
+        // - 1st candidate - input layer
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:446]: We should not requantize inputs I think, the only other solution (maybe even better) would be to take smallest scale factor of all inputs and apply it to make sure that no data is discarded (at cost of accuracy hit)

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 3]
```diff
@@ -454,16 +455,17 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
         static std::map<std::string, size_t> restarted_counter;
         auto restartedCountIt = restarted_counter.find(concatLayer->name);
         if (restartedCountIt == restarted_counter.end()) {
-            auto pos = restarted_counter.insert({ "concatLayer->name", 0 });
+            auto pos = restarted_counter.insert({ concatLayer->name, 0 });
             restartedCountIt = pos.first;
         }
 
-        if (restartedCountIt->second % 2 == 1) {
+        if (((restartedCountIt->second) / 2) % 2 == 1) {
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:462]: we should swap inputs every 2 iterations to make it possible to check if everything is fine after one swap

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp [hunk 4]
```diff
@@ -527,7 +529,7 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
                     gnalog() << "[UFS] from : " << concatLayer->name << " reached: " << layer->name;
                     // found that direct input to concat is a indirect parent of align filter - so no link required
                     auto info = LayerInfo(layer);
-                    if (!info.isWeightable() && !info.isActivation()) {
+                    if (!info.isWeightable() && !info.isActivation() && !info.isConst()) {
```
Comment (kmagiers) [inline inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:532]: requantizing const might be desirable, especially in cases that there are only input and concat layers.

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp (hunk 2 above)
Comment (bsochack) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:446]: one of the checks before is to check if all inputs have the same scale factor. Currently it is not supported a case when you have 2 inputs with 2 different scale factors. In general, the strategy that you are suggesting is the most reasonable.

**Comment by bsochack** [comment]
The change looks good. Just missing tests, but you are aware of it.

File: inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp (hunk 2 above)
Comment (kmagiers) [reply inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp:446]: without this check when there was "const" as first input and "input" as second input then sourceScaleFactor was set to be the const one. So the check was passing but input scale factor was not used.

**Review by bsochack — APPROVED** [review]

**Pull request merged**